Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcContentServer.cs

```diff
@@ -153,97 +156,174 @@ public async Task<GetStatsResponse> GetStatsAsync(GetStatsRequest request, Cance
         {
-            DateTime startTime = DateTime.UtcNow;
-
             try
             {
                 LogRequestHandling();
 
-                var cacheContext = new Context(new Guid(request.TraceId), _logger);
-
-                if (!_nameByDrive.TryGetValue(request.Drive, out string name))
-                {
-                    await responseStream.WriteAsync(
-                        new CopyFileResponse
-                        {
-                            Header = new ResponseHeader(startTime, false, (int)CopyFileResult.ResultCode.SourcePathError, $"'{request.Drive}' is an invalid cache."),
-                        });
-                }
+                // Get the content stream.
+                Context cacheContext = new Context(new Guid(request.TraceId), _logger);
+                HashType type = (HashType) request.HashType;
+                ContentHash hash = request.ContentHash.ToContentHash((HashType)request.HashType);
+                OpenStreamResult result = await GetFileStreamAsync(cacheContext, hash);
 
-                if (!_contentStoreByCacheName.TryGetValue(name, out var cache))
+                using (result.Stream)
                 {
-                    await responseStream.WriteAsync(
-                        new CopyFileResponse
-                        {
-                            Header = new ResponseHeader(startTime, false, (int)CopyFileResult.ResultCode.SourcePathError, $"'{name}' is an invalid cache name."),
-                        });
-                }
+                    // Figure out response headers.
+                    CopyCompression compression = CopyCompression.None;
+                    Metadata headers = new Metadata();
+                    headers.Add("ContentHash", hash.ToString());
+                    switch (result.Code)
+                    {
+                        case OpenStreamResult.ResultCode.ContentNotFound:
+                            headers.Add("Exception", "ContentNotFound");
+                            headers.Add("Message", $"Requested content {hash} not found.");
+                            break;
+                        case OpenStreamResult.ResultCode.Error:
+                            Debug.Assert(result.Exception != null);
+                            headers.Add("Exception", result.Exception.GetType().Name);
+                            headers.Add("Message", result.Exception.Message);
+                            break;
+                        case OpenStreamResult.ResultCode.Success:
+                            Debug.Assert(result.Stream != null);
+                            long size = result.Stream.Length;
+                            headers.Add("FileSize", size.ToString());
+                            if ((request.Compression == CopyCompression.Gzip) && (size > ContentStore.Grpc.CopyConstants.DefaultBufferSize)) {

```

Comment:
nit: inconsistent formatting (brace on the same line)

Reply:
Fixed